### PR TITLE
feat(call-omo-agent): support custom agents via dynamic resolution

### LINKED ISSUE #1706: [Feature]: [Feature Request] Support custom agents in call_omo_agent tool via ALLOWED_AGENTS whitelist
### Prerequisites

- [x] I will write this issue in English (see our [Language Policy](https://github.com/code-yeongyu/oh-my-opencode/blob/dev/CONTRIBUTING.md#language-policy))
- [x] I have searched existing issues and discussions to avoid duplicates
- [x] This feature request is specific to oh-my-opencode (not OpenCode core)
- [x] I have read the [documentation](https://github.com/code-yeongyu/oh-my-opencode#readme) or asked an AI coding agent with this project's GitHub URL loaded and couldn't find the answer

## Problem Description

The `call_omo_agent` tool uses a hardcoded `ALLOWED_AGENTS` whitelist to limit which agent types can be called. This prevents custom agents registered through the `customAgentSummaries` mechanism from being invoked by `call_omo_agent`.

While oh-my-opencode provides a `customAgentSummaries` mechanism that supports loading custom agents from the following sources:
- `configAgent`: from `config.agent`
- `userAgents`: from `~/.claude/agents/`
- `projectAgents`: from `.claude/agents/`
- `pluginAgents`: from plugins

These custom agents are only added to the `availableAgents` list for building Sisyphus's dynamic prompts, but they are **not** added to the `ALLOWED_AGENTS` whitelist.

## Expected Behavior

Custom agents registered through the `customAgentSummaries` mechanism should be callable by the `call_omo_agent` tool.

## Actual Behavior

The `call_omo_agent` tool checks whether `subagent_type` is in the `ALLOWED_AGENTS` whitelist during execution ([tools.ts:L52-L56](https://github.com/code-yeongyu/oh-my-opencode/blob/main/src/tools/call-omo-agent/tools.ts#L52-L56)). If not, it returns an error:

```typescript
if (![...ALLOWED_AGENTS].some(
  (name) => name.toLowerCase() === rawAgent.toLowerCase()
)) {
  return `Error: Invalid agent type "${rawAgent}". Only ${ALLOWED_AGENTS.join(", ")} are allowed.`
}
```

## Proposed Solutions

I suggest the following possible solutions (in order of priority):

### Solution 1: Dynamically generate ALLOWED_AGENTS (Recommended)

Modify the `createCallOmoAgent` function to dynamically merge custom agent names from `customAgentSummaries` into `ALLOWED_AGENTS`:

```typescript
export function createCallOmoAgent(
  ctx: PluginInput,
  backgroundManager: BackgroundManager
): ToolDefinition {
  const customAgentNames = ctx.customAgentSummaries
    ?.filter((a: any) => !a.hidden && !a.disabled)
    .map((a: any) => a.name.toLowerCase()) || []

  const allowedAgents = [...ALLOWED_AGENTS, ...customAgentNames]
  // ... use allowedAgents for validation
}
```

### Solution 2: Environment variable configuration

Add environment variable support in `constants.ts`:

```typescript
export const ALLOWED_AGENTS = [
  "explore",
  "librarian",
  "oracle",
  "hephaestus",
  "metis",
  "momus",
  "multimodal-looker",
  ...(process.env.OPENCODE_ADDITIONAL_AGENTS?.split(",") || []),
] as const
```

### Solution 3: Remove hardcoded restriction

Modify the validation logic to check if the agent exists in `config.agents` instead of just the hardcoded whitelist.

## Additional Information

- **Related code locations**:
  - `src/tools/call-omo-agent/constants.ts` - `ALLOWED_AGENTS` definition
  - `src/tools/call-omo-agent/tools.ts` - Validation logic
  - `src/agents/utils.ts:L351` - `customAgentSummaries` handling

- **Use case**:
  Users want to create custom multi-layer agents (e.g., multilayer agents: athena, hermes, ares, daedalus, argus, eris, proteus, themis) and use the `call_omo_agent` tool for task distribution and coordination.

### Alternatives Considered

_No response_

### Doctor Output (Optional)

```shell

```

### Additional Context

_No response_

### Feature Type

Configuration Option

### Contribution

- [ ] I'm willing to submit a PR for this feature
- [ ] I can help with testing
- [ ] I can help with documentation

### PR DESCRIPTION
**Pull request by brandonwebb-vista**
## Summary

Replace the hardcoded `ALLOWED_AGENTS` whitelist validation in `call_omo_agent` with dynamic agent resolution via `client.app.agents()`, and extend the agent loader to discover agents from all configuration paths.

Closes #1706

### Dynamic Agent Resolution (`call-omo-agent`)
- `agent-resolver.ts`: New `resolveCallableAgents()` queries registered agents at execute-time and **merges** them with `ALLOWED_AGENTS` (built-in agents are always callable, never dropped)
- `tools.ts`: Validate `subagent_type` against the dynamic set; fixed `subagent_type` parameter description
- `constants.ts`: Updated description to be path-agnostic (covers all loader paths, not just `~/.claude/agents/`)
- `tools.test.ts`: 18 tests covering dynamic resolution, merge behavior, fallback, and validation

### Agent Loader Extension (`claude-code-agent-loader`)
- `loader.ts`: Added `loadOpencodeGlobalAgents()` (`~/.config/opencode/agents/`) and `loadOpencodeProjectAgents()` (`<cwd>/.opencode/agents/`)
- `types.ts`: Extended `AgentScope` with `"opencode"` and `"opencode-project"`
- `loader.test.ts`: 15 tests covering YAML/MD parsing, tool extraction, scope assignment, and env-var paths

### Agent Config Handler Wiring (`agent-config-handler.ts`)
- Calls all 4 loaders: user, project, opencode-global, opencode-project
- OpenCode loaders run unconditionally (not gated by `claude_code.agents` config flag)
- Merge priority: `opencode-project > opencode-global > user > project > plugin`

## Test Results

- `bun test src/tools/call-omo-agent/`: **18 pass, 0 fail**
- `bun test src/features/claude-code-agent-loader/`: **15 pass, 0 fail**
- `bun run typecheck`: **clean**
- Full suite: **3540 pass, 5 fail** (all 5 are pre-existing failures in unrelated files)

## Agent Discovery Paths (after this PR)

| Path | Scope | Loader |
|------|-------|--------|
| `~/.claude/agents/` | `user` | `loadUserAgents()` |
| `<cwd>/.claude/agents/` | `project` | `loadProjectAgents()` |
| `~/.config/opencode/agents/` | `opencode` | `loadOpencodeGlobalAgents()` |
| `<cwd>/.opencode/agents/` | `opencode-project` | `loadOpencodeProjectAgents()` |

### CODE REVIEW

**Comment by brandonwebb-vista** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by brandonwebb-vista** [comment]
The CLA agent seems to be broken - it's looking at a "@brandon Webb" as the user (which is not my account) instead of "@brandonwebb-vista"

**Comment by alexmhtop** [comment]
> The CLA agent seems to be broken - it's looking at a "@brandon Webb" as the user (which is not my account) instead of "@brandonwebb-vista"

I think the first commit was done with the local git config which had another user than your GitHub one
image
Can you check it. Your feature is nice and I hope to see it implemented as well.

Thank you for the contribution! omo is amazing

**Comment by brandonwebb-vista** [comment]
@alexmhtop I realized that my email wasn't configured correctly in git global, so that's the reason for the difference. Short of recreating this MR, I'm hoping we can just force skip the CLA Assistant? It should be clear that I did sign the CLA

**Comment by brandonwebb-vista** [comment]
@code-yeongyu this MR should be ready to go

**Comment by tx1ee** [comment]
Awesome feature, expect code merge 👍

**Comment by brandonwebb-vista** [comment]
@code-yeongyu when can we expect to get this merged?

**Comment by tx1ee** [comment]
> @code-yeongyu when can we expect to get this merged?

Looking forward to this feature merger!!!

**Comment by code-yeongyu** [comment]
**Addressing remaining cubic findings from latest review:**

**P1 (agent-tool-restrictions.ts `?? {}` fallback):** This is the pre-existing default on `dev` branch — the PR diff for this file is whitespace-only. Changing to `EXPLORATION_AGENT_DENYLIST` breaks 3 existing tests (sisyphus, plan, prometheus need unrestricted access). Previous cubic review (round 3) explicitly requested keeping `{}` for OpenCode compatibility. Custom agents are user-created and intentionally unrestricted. Not a regression from this PR.

**P2 (merge order / canonical casing):** Built-in agents in `ALLOWED_AGENTS` are all lowercase — which IS their canonical form. Dynamic agents from the registry preserve original casing. Current order `[...dynamicAgents, ...ALLOWED_AGENTS]` prioritizes registry casing when overlapping, which is correct (registry is authoritative).

**P1 (throwOnError):** `client.app.agents()` is used identically in `delegate-task/subagent-resolver.ts` without `throwOnError`. The `try/catch` block already handles SDK errors gracefully with fallback to `ALLOWED_AGENTS`. Consistent with existing codebase patterns.

@cubic-dev-ai please re-review

**Comment by weijer** [comment]
Really need this feature. When can the current PR be merged

File: src/shared/agent-tool-restrictions.ts [hunk 1]
```diff
@@ -11,6 +11,7 @@ const EXPLORATION_AGENT_DENYLIST: Record<string, boolean> = {
   call_omo_agent: false,
```
[the remark this answers is not in the document]
Comment (brandonwebb-vista) [reply src/shared/agent-tool-restrictions.ts:13]: Fixed — restored `EXPLORATION_AGENT_DENYLIST` as the default fallback for unknown agents. Unknown/misspelled agent names now get the restrictive denylist (no write, edit, task, or call_omo_agent) instead of unrestricted access. Commit: aaba8618

File: src/shared/agent-tool-restrictions.ts [hunk 2]
```diff
@@ -47,7 +47,7 @@ const AGENT_RESTRICTIONS: Record<string, Record<string, boolean>> = {
 export function getAgentToolRestrictions(agentName: string): Record<string, boolean> {
   return AGENT_RESTRICTIONS[agentName]
     ?? Object.entries(AGENT_RESTRICTIONS).find(([key]) => key.toLowerCase() === agentName.toLowerCase())?.[1]
-    ?? {}
+    ?? EXPLORATION_AGENT_DENYLIST
```
[the remark this answers is not in the document]
Comment (brandonwebb-vista) [reply src/shared/agent-tool-restrictions.ts]: Fixed — `hasAgentToolRestrictions` now delegates to `getAgentToolRestrictions` so both functions use the same fallback logic. Unknown agents correctly return `true` for has-restrictions. Commit: 1e759ab6

**Comment by brandonwebb-vista** [comment]
@code-yeongyu can we merge this?

**Comment by tx1ee** [comment]
> @code-yeongyu can we merge this?

I am also waiting for this PR merger

**Comment by brandonwebb-vista** [comment]
@code-yeongyu can we merge? it's becoming a chore to continually come back and update/resolve merge conflicts

**Comment by brandonwebb-vista** [comment]
@code-yeongyu can we get this merged? Or perhaps you can open up OmA to collaborators?

**Comment by code-yeongyu** [comment]
[sisyphus-bot]

Hey @brandonwebb — first off, thank you so much for this PR! This is a really solid contribution. The dynamic agent resolution architecture is exactly the direction we wanted to go, and the test coverage (70+ new tests) is impressive. The precedence chain design is clean and the backward compatibility story is well thought out.

We ran two independent code reviews (security-focused + architecture-focused) and here is the consolidated feedback:

---

## Blocking Issues (must fix before merge)

### B1: Tool restriction bypass for custom agents (CRITICAL)
**File:** `src/shared/agent-tool-restrictions.ts:47-50`

The fallback to `EXPLORATION_AGENT_DENYLIST` means every custom agent is silently sandboxed to read-only mode — even if the user defines `tools: "Bash,Read,Write,Edit"` in their `.md` file. The `tools` config from agent definitions is never consulted at prompt time. Custom agents should either:
- Return `{}` (empty restrictions) when no matching agent is found in `AGENT_RESTRICTIONS`, letting the agent definition's `tools` config drive permissions, OR
- Explicitly document this as intentional sandboxing behavior

### B2: Prototype pollution inconsistency
**File:** `agent-definitions-loader.ts:70` (and `loader.ts:35,47,59,71`)

`opencode-config-agents-reader.ts:82` correctly uses `Object.create(null)` to prevent prototype pollution, but the other 4 accumulator sites use plain `{}`. A file named `__proto__.md` or frontmatter `name: __proto__` could hit the Object prototype. Apply `Object.create(null)` consistently across all agent-loading result objects.

### B3: Custom agent bash access undocumented
**File:** `agent-tool-restrictions.ts:50`

The `EXPLORATION_AGENT_DENYLIST` denies `write`, `edit`, `task`, `call_omo_agent` — but **not** `bash`. Project-scoped custom agents (from `.opencode/agents/` in a cloned repo) get shell execution. This may be intentional (matching Claude Code's trust model), but it must be explicitly documented or gated.

### B4: Two test failures in new code
**File:** `opencode-config-agents-reader.test.ts`

`"returns empty record when no opencode.json exists"` and `"handles malformed opencode.json gracefully"` — the function scans the global config dir too, which picks up real agents on the test machine. Mock `getOpenCodeConfigDir` to point at a temp dir.

---

## Non-blocking Issues (nice to have)

### N1: Path containment gap in opencode-config-agents-reader
`readOpencodeConfigAgents` reads `agent_definitions` paths with no containment check, unlike `plugin-config.ts` which uses `resolveAgentDefinitionPaths` with project boundary enforcement. A project's `opencode.json` could reference `../../other-project/agent.md`.

### N2: No caching in resolveCallableAgents
Called per tool invocation with no caching. A session-scoped cache would eliminate redundant SDK IPC calls.

### N3: parseToolsConfig duplicated 4x
Four near-identical implementations across loaders. Extract a single shared implementation into `shared/`.

### N4: agent_definitions path schema too permissive
`AgentDefinitionPathSchema = z.string()` — no min length or pattern validation. Empty strings resolve to cwd via `path.resolve(baseDir, "")`.

### N5: No test coverage for resolveAgentDefinitionPaths
Security-sensitive function (path traversal boundary) with zero tests. Edge cases: symlink escape, `..` traversal, absolute paths outside containment.

### N6: Mode type assertion without validation
`opencode-config-agents-reader.ts:68` — `mode: (agent.mode as "subagent" | "primary" | "all")` casts arbitrary user JSON strings without validation.

---

## Verdict

**Architecture: Strong.** The `client.app.agents()` → merge with `ALLOWED_AGENTS` design scales well. The 4-tier loader precedence (opencode-project > opencode-global > user > project > plugin) is correct. Config schema changes are backward compatible.

**Test coverage: Strong.** 70+ new tests with good given/when/then structure.

**Recommendation: Approve with required changes on B1–B4.** The non-blocking items can be addressed in follow-up PRs.

Again, really appreciate this contribution — the custom agents feature has been one of the most requested items. Looking forward to landing this! 🎉

**Comment by code-yeongyu** [comment]
[sisyphus-bot]

I went ahead and pushed fixes for all the review findings directly to this branch. Here's what was addressed:

**Blocking fixes:**
- **B1** (`agent-tool-restrictions.ts`): Changed fallback from `EXPLORATION_AGENT_DENYLIST` to `{}` — custom agents now get full tool access by default, matching Claude Code's trust model
- **B2** (`loader.ts`, `agent-definitions-loader.ts`): Applied `Object.create(null)` consistently across all 6 accumulator sites
- **B3** (`agent-tool-restrictions.ts`): Added documentation comment explaining bash access for custom agents
- **B4** (`opencode-config-agents-reader.test.ts`): Fixed 2 test failures by mocking `getOpenCodeConfigDir` to prevent global config leakage

**Non-blocking fixes:**
- **N1** (`opencode-config-agents-reader.ts`): Added `resolveAgentDefinitionPaths` containment check for `agent_definitions` paths
- **N2** (`agent-resolver.ts`): Added session-scoped `WeakMap` cache to `resolveCallableAgents`
- **N3**: Extracted shared `parseToolsConfig` into `src/shared/parse-tools-config.ts`, removed 4 duplicate implementations
- **N4** (`agent-definitions.ts`): Added `.min(1)` validation to `AgentDefinitionPathSchema`
- **N5**: Added `resolve-agent-definition-paths.test.ts` with 122 lines covering ~/ expansion, relative paths, boundary enforcement, and null containmentDir
- **N6** (`opencode-config-agents-reader.ts`): Added runtime validation for `mode` against allowed values

**Verification:**
- `bun run typecheck`: clean
- PR-related tests (136): all pass, 0 fail
- Full suite: 5530 pass, 11 fail (3 pre-existing, 8 flaky — all pass in isolation, confirmed test-order dependency unrelated to this PR)

The commit is `dd92ee0d`. Force-pushed to the branch.

**Comment by code-yeongyu** [comment]
[sisyphus-bot]

Pushed additional fix: replaced `mock.module()` with `spyOn()` in `opencode-config-agents-reader.test.ts` to prevent test pollution.

**Root cause:** `mock.module()` permanently replaced modules in bun's module cache, causing state leakage into `plugin-detection.test.ts`, `write-omo-config.test.ts`, and `config-loader.test.ts`.

**Result:**
- `bun run typecheck`: clean
- `bun test`: **5548 pass, 0 fail** (was 11 failures before)
- Rebased onto latest `dev`

Branch is ready for merge.
